design/01-proc-macro.md

```diff
@@ -0,0 +1,172 @@
+
+Scarb currently simply assumes a fixed, predefined plugin sets for `lib`, `starknet-contract` and `test` targets. This interface is not stable and is subject to changes in future compiler versions.
+
+### Plugins API in combination with Scarb
+
+Because of the use of shared reference to [`SyntaxGroup`], this API is hostile to attempts to externalize plugins into separate binaries/processes that communicate via some kind of RPC. The fact that AST structure is not stable does not help either. On the other hand, access to [`SyntaxGroup`] allows for powerful operations and is highly desired.
+
+The only feasible way of loading plugins is to load them directly into Scarb compiler's memory. This brings another problem though: Rust does not have a stable ABI, and thus such system would have to ensure that the plugin library was compiled with **exactly** the same Rust toolchain, for example `1.68.1-x86_64-pc-windows-msvc`.
+
+### Declaring package dependencies on plugins
+
+Scarb allows packages to declare dependency on a plugin through the same mechanism as on standard Cairo libraries. The plugin must be defined as a Scarb package defining `cairo-plugin` target in the package manifest.
+
+Since currently there is no mechanism to supply plugins implementation, it must be compiled into Scarb itself. Only `starknet` plugin is supplied this way.
+
+## Proposal
+
+This document proposes an introduction of new compiler plugins type, called procedural macros, as described below.
+
+### Plugins will be compiled into shared libraries
+
+- Plugins will be implemented in Rust.
+- Plugins will be compiled into shared libraries, that Scarb can load at the runtime through dynamic loading mechanism of the host ope rating system.
+- Plugins will be compiled to `.dylib` files on OSX, `.so` files on Linux and `.dll` on Windows.
+- Scarb will use [`libloading` crate](https://docs.rs/libloading/latest/libloading/), which exposes a cross-platform interface for interactions with platform dynamic library loading primitives, to load shared libraries into memory.
+
+### Plugins will be special Scarb packages containing Rust code
+
+- A plugin will be a special Scarb package, which will use a new native target type: `cairo-plugin`.
+- The plugin package will contain Rust plugin source code, which will be shipped to users on Scarb project resolution.
+- The plugin source code will be compiled on Scarb users system only.
+- Enabling this target will mean that this package does not contain any Cairo code.
+- This target will be *exclusive*:
+    - It will block defining other targets for the package, not even `lib`.
+    - It will also not be possible to declare dependencies, or specify Cairo compiler settings, it won't make sense for these packages.
+- Scarb will automatically generate a `Cargo.toml` manifest file, based on contents of `Scarb.toml` manifest. The Cargo manifest will be used for the plugin compilation.
```

Comment:
This does not tell _when_ `Cargo.toml` will be generated. I think it should be part of workspace resolution (i.e. `scarb fetch`). Furthermore, I believe calling `cargo fetch` for all plugins in the workspace should also be done during this phase.